xla/backends/cpu/runtime/onednn/onednn_thunk.cc

```diff
@@ -0,0 +1,228 @@
+    BufferAllocation::Slice& slice = op_buffers_.results_buffers[i];
+    TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase res,
+                        params.buffer_allocations->GetDeviceAddress(slice));
+    results.emplace_back(res);
+    auto memref =
+        CreateMemrefFromShape(op_buffers_.results_shapes[i], res.opaque());
+    runtime->resources.result_memrefs.push_back(std::move(memref));
+    ABSL_ANNOTATE_MEMORY_IS_INITIALIZED(results[i].opaque(), results[i].size());
+    VLOG(3) << absl::StreamFormat("  res: %s in slice %s (%p)",
+                                  op_buffers_.results_shapes[i].ToString(true),
+                                  slice.ToString(), results[i].opaque());
+  }
+
+  // Borrow the FFI call frame from the object pool and update with the actual
+  // device memory addresses.
+  TF_ASSIGN_OR_RETURN(auto call_frame, call_frames_.GetOrCreate());
+  TF_RETURN_IF_ERROR(call_frame->UpdateWithBuffers(arguments, results));
+
+  // Do a heap allocation of the ExecutionContext, to ensure that the context
+  // remains valid until the execution is complete.
+  auto owned_context = std::make_unique<ffi::ExecutionContext>();
+
+  // Insert the oneDNN-related objects into the execution context. This allows
+  // the FFI handler to safely access the oneDNN engine, stream, and resources
+  // until the execution of the custom call is complete.
+  auto status = owned_context->Insert(&runtime->cpu_engine);
+  if (!status.ok()) {
+    return Internal("Failed to add oneDNN engine to the execution context: %s",
+                    status.message());
+  }
+  status = owned_context->Insert(&runtime->onednn_stream);
+  if (!status.ok()) {
+    return Internal("Failed to add oneDNN stream to the execution context: %s",
+                    status.message());
+  }
+  status = owned_context->Insert(runtime->threadpool.get());
+  if (!status.ok()) {
+    return Internal(
+        "Failed to add oneDNN threadpool to the execution context: %s",
+        status.message());
+  }
+  status = owned_context->Insert(&runtime->resources);
+  if (!status.ok()) {
+    return Internal(
+        "Failed to add oneDNN resources to the execution context: %s",
+        status.message());
+  }
+
+  ffi::CallOptions call_options = {
+      custom_call_params->run_id,
+      custom_call_params->device_ordinal,
+      ffi::CallOptions::CpuOptions{custom_call_params->intra_op_thread_pool},
+      /*called_computation=*/nullptr,
+      owned_context.get(),
+      execution_state_.get()};
+
+  ffi::HandlerRegistration& handler = std::get<1>(target_);
+
+  auto executed =
+      ffi::CallAsync(handler.bundle.execute, *call_frame, call_options);
+
+  executed.AndThen(
+      [runtime = std::move(runtime), owned_context = std::move(owned_context)] {
+        // runtime and context_holder will be automatically destroyed here
+      });
+
+  return executed;
+}
+
+OneDnnThunk::BufferUses OneDnnThunk::buffer_uses() const {
```

Comment:
Nit: We can use the one from `CustomCallThunk`.

Reply:
Done.